From 447200eb56262fb3337990f336e081a3283b7dd0 Mon Sep 17 00:00:00 2001
From: Jan Nikitenko <jan.nikitenko@gmail.com>
Date: Fri, 28 Nov 2008 08:52:58 +0100
Subject: [PATCH 033/117] MIPS: Fix oops in dma_unmap_page on not coherent mips platforms

commit 447200eb56262fb3337990f336e081a3283b7dd0 from linux-mips

dma_cache_wback_inv() expects virtual address, but physical was provided
due to translation via plat_dma_addr_to_phys().
If replaced with dma_addr_to_virt(), page fault oops from dma_unmap_page()
is gone on au1550 platform.

Signed-off-by: Jan Nikitenko <jan.nikitenko@gmail.com>
Acked-by: Atsushi Nemoto <anemo@mba.ocn.ne.jp>
Signed-off-by: Ralf Baechle <ralf@linux-mips.org>
(cherry picked from commit 2d3f83c7afa9457c25fd7968da700ebecca9dd85)
---
 arch/mips/mm/dma-default.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/arch/mips/mm/dma-default.c b/arch/mips/mm/dma-default.c
index e6708b3..1a58be3 100644
--- a/arch/mips/mm/dma-default.c
+++ b/arch/mips/mm/dma-default.c
@@ -225,7 +225,7 @@ void dma_unmap_page(struct device *dev, dma_addr_t dma_address, size_t size,
 	if (!plat_device_is_coherent(dev) && direction != DMA_TO_DEVICE) {
 		unsigned long addr;
 
-		addr = plat_dma_addr_to_phys(dma_address);
+		addr = dma_addr_to_virt(dma_address);
 		dma_cache_wback_inv(addr, size);
 	}
 
-- 
1.7.0

